plugins/modules/sensu_go_asset.py

```diff
@@ -65,68 +71,57 @@
       sensio.io.bonsai.tier: Community
       sensio.io.bonsai.version: 4.0.0
       sensio.io.bonsai.tags: ruby-runtime-2.4.4
-'''
+"""
 
-RETURN = '''
-'''
+RETURN = """
+"""
 
-from ansible.module_utils.basic import AnsibleModule
-from ansible_collections.sensu.sensu_go.plugins.module_utils.base import SensuObject
-
-
-class SensuAsset(SensuObject):
-    def __init__(self, module):
-        super(SensuAsset, self).__init__(module)
+import operator
 
-        self.path = '/assets/{0}'.format(self.params['name'])
+from ansible.module_utils.basic import AnsibleModule
 
-        for key in (
-            'download_url',
-            "sha512",
-            "filters",
-            "headers"
-        ):
-            if self.params[key] is not None:
-                if key == "download_url":
-                    self.payload["url"] = self.params[key]
-                else:
-                    self.payload[key] = self.params[key]
+from ansible_collections.sensu.sensu_go.plugins.module_utils import (
+    arguments, errors, utils,
+)
 
 
 def main():
-    argspec = SensuAsset.argument_spec()
-    argspec.update(
-        dict(
-            state=dict(
-                default='present',
-                choices=['present'],
-            ),
+    module = AnsibleModule(
+        supports_check_mode=True,
+        argument_spec=dict(
+            arguments.MUTATION_ARGUMENTS,
             download_url=dict(
                 required=True,
             ),
             sha512=dict(
                 required=True,
             ),
             filters=dict(
-                type='list',
+                type="list",
                 default=[],
             ),
             headers=dict(
-                type='dict',
+                type="dict",
                 default={},
             ),
-        )
+        ),
     )
 
-    module = AnsibleModule(
-        supports_check_mode=True,
-        argument_spec=argspec,
+    client = arguments.get_sensu_client(module.params)
+    path = "/assets/{0}".format(module.params["name"])
+    payload = arguments.get_mutation_payload(
+        module.params, "download_url", "sha512", "filters", "headers",
     )
+    payload["url"] = payload.pop("download_url")  # Remap download_url -> url
```

Comment:
UPDATE: I'm passing my idea even if we don't implement it yet, so that we can quickly discuss. I think it would be cleanest if we introduce a "mapper" class which does two things:

a) constructs itself either out of Ansible module params or out of Sensu API JSON payload
b) is able to perform comparison between two instances of itself

Namely:

```python
class Asset:
  @classmethod
  def from_ansible(cls, ansible_args):
    cls(...)

  @classmethod
  def from_sensu(cls, sensu_args):
    cls(...)

  def __eq__(self, other):
    pass
```

Having all the mapping shit delegated to such a class things would get nice. Here's how we would use it:

```python
client = arguments.get_sensu_client(module.params)
path = "/assets/{0}".format(module.params["name"])
desired = Asset.from_ansible(module.params)  # This line here

try:
	changed, asset = utils.sync(
		module.params["state"], client, path, desired, module.check_mode,
	)
	module.exit_json(changed=changed, asset=asset)
except errors.Error as e:
	module.fail_json(msg=str(e))
```

`sync` method would then feed GET-ed JSON to the `Asset.from_sensu()` and then use `==` to find out whether desired and remote are the same.

@tadeboro looking forward to your opinion. With Aljaž we were working quite some time on the idea, even did some POCs and it seemed to simplify things.

Reply:
For modules that need to perform complex mapping, extracting all that logic into a separate function should clear the `main` method up. I did not do the extraction of the functionality in this case because the transformation is almost trivial.

As for the `Asset` kind of classes, I would find this kind of functionality useful if the payload and the Sensu's responses can have a different structure. If the payload and the server's answer always have the same general structure, that would be an overkill.

And even if the comparison gets a bit more complex, I would probably first try to add a `comparator` parameter to the `sync` method that would take a callable that can compare payload and server's response.

Reply:
Lets merge as-is for now. I will still ask Aljaž to introduce the change because I believe with correctly designed mapper base class the code readability will become awesome (even if it's great already with this PR).